Add an air cushion surge tank model

pandeysudan27 opened the pull request.

pandeysudan27 committed:
added a dummy change

pandeysudan27 committed:
Revert "added a dummy change"

This reverts commit adfcfb3ec0375fcea791f15f2e4283cda37a1530.

pandeysudan27 committed:
Auto stash before rebase of "componentBranch"

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,79 @@
+  Modelica.SIunits.Force F_p "pressure force";
+  Modelica.SIunits.Force F_f "friction force";
+  Modelica.SIunits.Force F_g "friction force";
+  Modelica.SIunits.Pressure p_c "air-cushion force";
```
Convention is that description strings should start with an upper case. E.g., `"Water mass"`.

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,79 @@
+within OpenHPL.Waterway;
+model AirCushionSurgeTank "Model of air-cushion surge tank"
+  outer Parameters para "using standard class with constants";
```
should be parameters in the description string.

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,79 @@
+ annotation (
+    Documentation(info="<html>
+<p>The simple model of the air cushion surge tank, which described by the momentum and mass differential equations. The mass balance depends on inlet and outlet mass flow rates. The momentum balance depends on inlet momentum and pressure dorp through the surge pipe together with gravity and friction forces. The main defined variable are <i>V_dot_s</i> and <i>h_s&nbsp;</i>(the flow rate and water  ...
+<p><img src=\"modelica://OpenHPL/Resources/Images/surgepic.png\"/></p>
```
Could we start wit the SVGs instead of pngs right away?

pandeysudan27 replied:
Absolutely, but i have problems with _arrow_  not showing in right direction in svg of inkscape. i mean the vertically downward arrows shows horizontally in the inkscape.

dietmarw replied:
Could you add the svg too. So I can have a look at it.

pandeysudan27 replied:
ACST.svg is added in the Resource/Images that depict image for air cushion surge tank.

dietmarw replied:
So I checked and it looks fine in Inkscape:
![image](https://user-images.githubusercontent.com/9332/69441200-8f3b7180-0d4a-11ea-9da7-14993b28fd6e.png)

What do you mean does not look right?

dietmarw replied:
Btw some lines are not aligned ;-)

pandeysudan27 replied:
It consists of a mesh grid color for water and air. The svg image is not rendered (for example by the browser) for the mesh grid. But the problem in arrowhead is not occurring.

dietmarw replied:
Also maybe the colours should be matching.

pandeysudan27 replied:
Sorry, I had arrowhead problems with different figures. But .svg image consisting of colors in mesh grid when opened by the browser I cannot see the color. The image looks like this
![2019-11-22_172811](https://user-images.githubusercontent.com/24632158/69442931-b182be80-0d4d-11ea-86bf-c3565d729d9c.png)

dietmarw replied:
I've posted a solution video in the Teams channel

pandeysudan27 committed:
Model of air cushion surge tank added.

The air-cushion surge tank model is added with default value of geometrical parameters from Torpa HPP, Norway.

pandeysudan27 committed:
Simple typo fixed.

pandeysudan27 committed:
ACST.svg file added for review in inkscape

pandeysudan27 commented on OpenHPL/Tests/package.order:
```diff
@@ -1,5 +1,6 @@
 Fittings
 HPSimplePenstock
+HPAirCushionSurgeTank
 HPSimplePenstockWithoutSurge
```
Test file for air cushion surge tank started.

dietmarw commented:
You should also revert the "dummy change". Simply right-click and revert commit.

dietmarw commented on OpenHPL/Tests/package.order:
```diff
@@ -61,3 +62,4 @@ HydroCordModelKPDraftTube
 HydroCordModelFrancisDraftTube
 HydroCordModeKPlFrancisDraftTube
 HydroCordModelDraftTube3
+HPSimple_FranciSurc
```
This one does not exist in the library and hence OM will not load

dietmarw committed:
renamed to give more elaborate name

dietmarw committed:
Fix package.order (removing non-existing model)

dietmarw committed:
Start on SVG fix.

pandeysudan27 committed:
Air cushion surge tank svg figure added. The problem of mesh grid and arrow was solved

pandeysudan27 committed:
Method abs(v) was changed to max(-v,v) for simulation speed

Some testing was done with abs and max method in modelica. It shows that max is 13 times faster than abs.

pandeysudan27 committed:
Revert "Method abs(v) was changed to max(-v,v) for simulation speed"

This reverts commit c9a58744a95cce3017f46ee7daaeece7bdf7fed5.

dietmarw commented on OpenHPL/Functions/DarcyFriction/Friction.mo:
```diff
@@ -15,9 +15,9 @@ protected
 algorithm
-  N_Re := rho * abs(v) * D / mu;
+  N_Re := rho * max(-v,v) * D / mu;
   f := fDarcy(N_Re, D, eps);
```
Sorry, that is really not good.  This change should be reverted. If this really is a performance issue (could you give  a test case) then this is a tool issue and should not be solved in the library with work arounds.

pandeysudan27 replied:
When i ran stand alone surge tank model i checked it in openmodelica with and without max method. How shlould i create a test case for this in OpenHPL. Or I will revert it now and maybe we fix it in future.

dietmarw replied:
In that case that would definitely be a bug in OpenModelica. So good if there is a small test case so we can run a comparison also with other tools.

dietmarw commented:
The dimensions of the SVG need to get fixed again.

dietmarw commented:
Also the tips of the arrows should not overlap the lines they are pointing to.

pandeysudan27 committed:
svg and png for air cushion surge tank added

pandeysudan27 committed:
svg file resized to page

dietmarw commented:
This looks already quite good now. I have a couple of additional comments. Bear with be for being this pedantic but this just helps avoiding extra work later :-)

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,80 @@
+  F_g = m * para.g * cos_theta;
+ annotation (
+    Documentation(info= "<html>
+<p>The simple model of the air-cushion surge tank, which is described by the momentum and mass balance differential equations. The mass balance depends on inlet and outlet mass flow rates. The momentum balance depends on inlet momentum and pressure dorp through the surge pipe together with gravity and friction forces. The main defined variable are <i>V_dot</i> and <i>h  </i>(the flow rate and wat ...
```
When revering to code variables you should use `<code>X_y</code>` instead of `<i>`. Actually `<i>` should not be used at all. If you want to emphasise use `<em>` instead.

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,80 @@
+    Documentation(info= "<html>
+<p>The simple model of the air-cushion surge tank, which is described by the momentum and mass balance differential equations. The mass balance depends on inlet and outlet mass flow rates. The momentum balance depends on inlet momentum and pressure dorp through the surge pipe together with gravity and friction forces. The main defined variable are <i>V_dot</i> and <i>h  </i>(the flow rate and wat ...
+<p><img src=\"modelica://OpenHPL/Resources/Images/AirCushionSurgeTank.png\" style=\"width:100%\"/></p>
+<p>The simple surge tank model can be found in this link:&nbsp;<a href=\"http://www.ep.liu.se/ecp/article.asp?article=049&issue=138&volume=\">http://www.ep.liu.se/ecp/article.asp?article=049&amp;issue=138&amp;volume=</a></p>
```
Instead of linking to explicit documents here add the paper under `OpenHPL.UsersGuide.References` as a new entry and link to that class with the name of the reference here..

dietmarw reviewed (commented):
In general, looking at the name. It would be probably better to call it `SurgeTankAirCushion`. This also looks better in the Model listing. I would then also rename the image files accordingly.

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,80 @@
+  Modelica.SIunits.Force F "Total force acting in the surge tank";
+  Modelica.SIunits.Area A = (pi*D ^ 2) / 4 "Cross sectional area of the surge tank";
+  Modelica.SIunits.Length l = h / cos_theta "Length of water in the surge tank";
+  Real cos_theta = H / L "slope ratio";
```
```suggestion
  Real cos_theta = H / L "Slope ratio";
```

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,80 @@
+    //T_n = T_i;
+  end if;
+equation
+  //// mass and momentum balance
```
```suggestion
```

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,80 @@
+  end if;
+equation
+  //// mass and momentum balance
+  der(m) = m_dot;
```
```suggestion
  der(m) = m_dot "Mass balance";
```

dietmarw commented on OpenHPL/Waterway/AirCushionSurgeTank.mo:
```diff
@@ -0,0 +1,80 @@
+equation
+  //// mass and momentum balance
+  der(m) = m_dot;
+  der(M) = M_dot+F;
```
```suggestion
  der(M) = M_dot+F "Momentum Balance";
```

pandeysudan27 committed:
The last suggestions were implemented.

dietmarw commented:
Closes in favour of #14
